File: sdk/python3/nmidsdk/client/discovery.py
import json
import random
import etcd3
import nmidsdk.model.const as const
from threading import Thread
from nmidsdk.client.client import Client

class Consumer:
    def __init__(self, etcd_host, etcd_port, username="", password=""):
        self.etcd_host = etcd_host
        self.etcd_port = etcd_port
        self.username = username
        self.password = password
        self.etcd_cli = None
        self.workers = {}

    def etcd_client(self):
        self.etcd_cli = etcd3.client(host="192.168.10.195", port=2379, user="root", password="123456", cert_key=None, cert_cert=None)
        return self.etcd_cli
    
    def etcd_watch(self):
        if self.etcd_cli is not None:
            watch = self.etcd_cli.watch_prefix(const.EtcdBaseKey)
            Thread(target=self._watch_loop, args=(watch,)).start()

    def _watch_loop(self, watch):
        for evnet in watch:
            self.workers = self.get_all_worker_ins(const.EtcdBaseKey)

    def get_all_worker_ins(self, prefix):
        all_worker_ins = {}
        try:
            resp = self.etcd_cli.get_prefix(prefix)
            for kv in resp:
                worker_ins = json.loads(kv.value.decode())
                all_worker_ins.update(worker_ins)
        except Exception as e:
            print(f"Error getting data from etcd: {e}")
            return None
        return all_worker_ins
    
    def discovery(self, func_name):
        if addrs := self.workers.get(func_name):
            index = random.randint(1, 10000) % len(addrs)
            nmid_addr = addrs[index]
            parts = nmid_addr.split(':')
            SERVER_HOST, SERVER_PORT = parts[0], int(parts[1])
            return Client("tcp", (SERVER_HOST,SERVER_PORT))
        
        return None

File: pkg/client/discovery.go
package client

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/HughNian/nmid/pkg/model"
	"github.com/HughNian/nmid/pkg/utils"
	clientv3 "go.etcd.io/etcd/client/v3"
)

type Consumer struct {
	EtcdAddrs []string
	Username  string
	Password  string
	EtcdCli   *clientv3.Client
	Workers   map[string][]string
}

func (c *Consumer) EtcdClient() *clientv3.Client {
	v3Config := clientv3.Config{
		Endpoints:   c.EtcdAddrs,
		DialTimeout: 5 * time.Second,
	}
	if c.Username != "" && c.Password != "" {
		v3Config.Username = c.Username
		v3Config.Password = c.Password
	}

	cli, err := clientv3.New(v3Config)

	if err != nil {
		// handle error!
		fmt.Printf("connect to etcd failed, err:%v\n", err)
		return nil
	}

	return cli
}

func (c *Consumer) EtcdWatch() {
	watchCh := c.EtcdCli.Watch(context.TODO(), model.EtcdBaseKey, clientv3.WithPrefix())
	go func() {
		for {
			<-watchCh

			c.Workers = c.GetAllWorkerIns(model.EtcdBaseKey)

			//do prometheus discovery count
			discoveryCount.Inc("etcd")
		}
	}()
}

func (c *Consumer) Discovery(funcName string) (client *Client) {
	if addrs, ok := c.Workers[funcName]; ok {
		index := utils.RandomInt(1, 10000) % len(addrs) //随机负载，todo使用更多的loadbalnce算法
		nmidAddr := addrs[index]

		client = NewClient("tcp", nmidAddr)
		return client
	}

	return
}

func (c *Consumer) GetAllWorkerIns(prefix string) map[string][]string {
	var allWorkerIns = make(map[string][]string)

	//use key prefix get the workerins
	resp, err := c.EtcdCli.Get(context.Background(), prefix, clientv3.WithPrefix())
	if err != nil {
		return nil
	}

	for i := range resp.Kvs {
		if v := resp.Kvs[i].Value; v != nil {
			// fmt.Printf("%s-%s\n", resp.Kvs[i].Key, resp.Kvs[i].Value)

			var workerIns = make(map[string][]string)
			ret := json.Unmarshal(resp.Kvs[i].Value, &workerIns)
			if ret == nil {
				for key, val := range workerIns {
					allWorkerIns[key] = val
				}
			}
		}
	}

	return allWorkerIns
}
